### PR TITLE
fix(ngMock/$httpBackend): correctly ignore query params in {expect,when}Route

### PR DESCRIPTION
**Pull request by thorn0**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

fix

**What is the current behavior? (You can also link to an open issue here)**

#14173

**Does this PR introduce a breaking change?**

no

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [x] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

### CODE REVIEW

**Review by gkalpak — COMMENTED** [review]
Actually, I think it would be better to re-use the method used in `ngRoute` (see https://github.com/angular/angular.js/issues/14173#issuecomment-192649766).
It would be similar to how we re-use [shallowCopy()][1] (configured in [angularFiles.js][2]).

[1]: https://github.com/angular/angular.js/blob/07849779ba365f371a8caa3b58e23f677cfdc5ad/src/shallowCopy.js
[2]: https://github.com/angular/angular.js/blob/73c6467f1468353215dc689c019ed83aa4993c77/angularFiles.js

**Comment by gkalpak** [comment]
I would put it in `src/` (since it will be used in different modules).

**Comment by thorn0** [comment]
@gkalpak PTAL

File: src/routeToRegExp.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * @return {?Object}
+ *
+ * @description
+ * Normalizes the given path, returning a regular expression
+ * and the original path.
+ *
+ * Inspired by pathRexp in visionmedia/express/lib/utils.js.
+ */
+function routeToRegExp(path, opts) {
+  if (!angular.isString(path)) path = '';
```
Comment (gkalpak) [inline src/routeToRegExp.js]: AFAICT, this function would never be called with something that is not a string (if people followed the docs :grin:).
I suggest we remove this line (and fail hard if `path` is not a string, so that we can detect issues quicker), instead of covering it up with a route that does not correspond to the passed in `path`.

File: test/ngMock/angular-mocksSpec.js [hunk 2]
```diff
@@ -1943,10 +1943,30 @@ describe('ngMock', function() {
         );
         they('should ignore query param when matching in ' + routeShortcut + ' $prop method', methods,
           function() {
-            hb[routeShortcut](this, '/route/:id').respond('path');
+            var paramsSpy = jasmine.createSpy('paramsSpy');
+            hb[routeShortcut](this, '/route/:id').respond(function(method, url, data, headers, params) {
+              paramsSpy(params);
+              // status, response, headers, statusText, xhrStatus
+              return [200, 'path', { 'x-header': 'foo' }, 'OK', 'complete'];
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: I don't think `complete` is used (here and below).

File: test/ngMock/angular-mocksSpec.js [hunk 3]
```diff
@@ -1943,10 +1943,30 @@ describe('ngMock', function() {
         );
         they('should ignore query param when matching in ' + routeShortcut + ' $prop method', methods,
           function() {
-            hb[routeShortcut](this, '/route/:id').respond('path');
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: Why the changes in this test? What are you trying to verify?

File: test/ngMock/angular-mocksSpec.js [hunk 4]
```diff
@@ -1943,10 +1943,30 @@ describe('ngMock', function() {
...
             hb.flush();
-            expect(callback).toHaveBeenCalledOnceWith(200, 'path', '', '', 'complete');
+            expect(callback).toHaveBeenCalledOnceWith(200, 'path', 'x-header: foo', 'OK', 'complete');
+            expect(paramsSpy).toHaveBeenCalledOnceWith({id: '123', q: 'str', foo: 'bar'});
+          }
+        );
+        they('should ignore query param when matching eager parameters in ' + routeShortcut + ' $prop method', methods,
+          function() {
+            var paramsSpy = jasmine.createSpy('paramsSpy');
+            hb[routeShortcut](this, '/route/:id*').respond(function(method, url, data, headers, params) {
```
Comment (gkalpak) [inline test/ngMock/angular-mocksSpec.js]: How about adding a test without the `*` (for good measure)?

**Review by gkalpak — CHANGES_REQUESTED** [review]
Looks great 👍 
Just a couple of minor things.

File: src/ngMock/angular-mocks.js [hunk 5]
```diff
@@ -1481,8 +1483,9 @@ function createHttpBackendMock($rootScope, $timeout, $delegate, $browser) {
    *      ```
    *    – The respond method takes a set of static data to be returned or a function that can
    *    return an array containing response status (number), response data (Array|Object|string),
-   *    response headers (Object), and the text for the status (string). The respond method returns
-   *    the `requestHandler` object for possible overrides.
+   *    response headers (Object), HTTP status text (string), and XMLHttpRequest status (string:
```
[the remark this answers is not in the document]
Comment (thorn0) [reply src/ngMock/angular-mocks.js]: It doesn't get it. It returns it. See http://jsbin.com/vumefiz/edit?html,js,output

File: test/ngMock/angular-mocksSpec.js (hunk 2 above)
Comment (thorn0) [reply test/ngMock/angular-mocksSpec.js]: I'm verifying that `params` are correct. Without the fix, `paramSpy` is called with `{id: '123?q=str&foo=bar', q: 'str', foo: 'bar'}`.

File: src/ngMock/angular-mocks.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: Sorry, I misread that 😳

[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: Actually, our [createResponse()][1] method isn't consistently treating `XHR status`. It does not accept one when passing the data directly to `.respond()` (and it automatically uses `complete`as the XHR status), but it does accept one when a function is passed to `createResponse()`.

I think this is a bug (or at least unintentional). `createResponse()` should look something like:

```diff
 function createResponse(status, data, headers, statusText) {
-  if (angular.isFunction(status)) return status;
+  if (angular.isFunction(status)) {
+    return function(method, url, data, headers, params) {
+      var response = status(method, url, data, headers, params);
+      response[4] = 'complete';
+      return response;
+    };
+  }

   return function() {
     return angular.isNumber(status)
         ? [status, data, headers, statusText, 'complete']
         : [200, status, data, headers, 'complete'];
   };
 }
```

But this is unrelated to this PR and would probably be a breaking change now 😞 

[1]: https://github.com/thorn0/angular.js/blob/04024ed13437b6e8ffca2158625f006ee6cfc8cc/src/ngMock/angular-mocks.js#L1355-L1363

[the remark this answers is not in the document]
Comment (thorn0) [reply src/ngMock/angular-mocks.js]: Why do you think so? What if we want to test the behavior of our app with XHR statuses other than `complete` (e.g. `error`)?

[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: Because the XHR status is not something that can be arbitrarily specified. It is tied to the way the XHR complete. You can't have `error` with a "normal" response (it is something that will not come up in the actual app).

[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: I.e., the only valid XHR status when returning a response (via the function returned by `createResponse()`) is `complete`.

[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: But (like I said), this is not something to worry about in this PR. It is not related (and that ship has probably sailed :grin:).

[the remark this answers is not in the document]
Comment (thorn0) [reply src/ngMock/angular-mocks.js]: > You can't have error with a "normal" response (it is something that will not come up in the actual app

I believe I can. It happens in case of network error, doesn't it?

[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: Yes, that's not a "normal" response. I.e. an XHR status other than `complete` can only happen if the XHR's `.onabort`/`.onerror`/`.ontimeout` is called. In that case, you would have a status of `0` and empty values for other fields (`statusText`, `headers`, etc).

But the mock `$httpBackend` is not really designed to let you test these scenarios. Sure you can emulate it by returning the correct values from your response function, you need to know `$httpBackend` internals. E.g. returning `[200, 'OK', ..., 'aborted']` is a scenario that will never happen in the actual app and will lead to unexpected results (because `aborted` will have been called with a "success" status).

So, ideally, we should not let users specify the XHR status, but instead infer it from other params.
Even more ideally, `$httpBackend` could support testing scenarions when an XHR errored, or timed out or was aborted (and again automatically setting the correct XHR status; not letting the user specify it).

But as it is now, `$httpBackend` is better used with "normal" responses (unless you really know what you are doing) 😛

[the remark this answers is not in the document]
Comment (thorn0) [reply src/ngMock/angular-mocks.js]: In a separate PR, we can add simple checks and throw in cases like `[200, 'OK', ..., 'aborted']`. What are your thoughts on that?

[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ngMock/angular-mocks.js]: I wouldn't be opposed to that :smiley:

File: test/ngMock/angular-mocksSpec.js (hunk 4 above)
Comment (thorn0) [reply test/ngMock/angular-mocksSpec.js]: I misread this question before. There already is a test without `*`. I added a test with `*`. I thought you were asking about a test without any routing parameters at all.

Comment (gkalpak) [reply test/ngMock/angular-mocksSpec.js]: 🙈

**Comment by thorn0** [comment]
@gkalpak Should the URL `/route/123/` match the route `/route/:id`?

**Comment by thorn0** [comment]
The current behavior is a bit unexpected:
```js
ret.regexp = new RegExp('^' + url, 'i');
```
Which means even `/route/123/bla/bla/bla` matches the route `/route/:id`. Do we really want to keep it?

**Comment by gkalpak** [comment]
No, I meant the behavior wrt trailing `/`. Matching URLs both with and without a trailing `/` is reasonable (and is also how `$route` works by default - which is unrelated, but still :grin:).

Matching the `bla/bla/bla` part is a bug imo. If one what to do that, they can use a `:rest*?` segment at the end.

So, I woud vote for matching with and without trailing `/`, but not match with trailing segments (e.g. `bla/bla/bla`). WDYT?

**Comment by thorn0** [comment]
Okay. Matching the `bla/bla/bla` part isn't documented, so we can consider it a bug.

**Comment by thorn0** [comment]
@gkalpak PTAL. Travis's fail is irrelevant.

**Review by gkalpak — APPROVED** [review]

**Comment by gkalpak** [comment]
Thx, @thorn0 ❤️
(Added some more info in the commit message and merged.)
